File: 050 Pow(x, n)/pow_recursive.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            x = 1 / x
            n = -n
        return self.fast_pow(x, n)
        
    
    def fast_pow(self, x, n):
        if n == 0:
            return 1.0
        half = self.fast_pow(x, n // 2)
        if n % 2 == 0:
            return half * half
        else:
            return half * half * x

File: 050 Pow(x, n)/Pow_Recursive.java
class Solution {
    public double myPow(double x, int n) {
        long N = n;
        if (N == 0) {
            return 1.0;
        }
        
        if (N < 0) {
            N = -N;
            x = 1 / x;
        }
        
        double half = myPow(x, (int) (N / 2));
        if (N % 2 == 0) {
            return half * half;
        }
        return half * half * x;
    }
}